dali/kernels/imgproc/convolution/separable_convolution_cpu.h

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_SEPARABLE_CONVOLUTION_CPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_SEPARABLE_CONVOLUTION_CPU_H_
+
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/convolution_cpu.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Apply convolution in all axes, starting from the innermost to outermost.
+ *        If channel axis is pressent, the convolution is not applied there.
+ *
+ * If `Out` is same as `W` the intermediate stages are written to out,
+ * otherwise the temporary buffer of `W` is required and allocated in scratchpad
+ *
+ * `W` is currently assumed to be float
+ *
+ * `windows` and `scales` are specified per axis.
+ *
+ * Specialized for 1, 2 or 3 axes, to not go overboard with TMP for generic solutions
+ *
+ * TODO(klecki): For more dimension, fusing a permute step when writing the result
+ * could allow for processing all steps with innermost, contiguous dimension.
+ * For example DHWC->DWHC->HWDC->DHWC, while applying convolutions for W, H, D respectively.
```

Comment:
I wouldn't mark it as TODO - it doesn't seem like a good idea to transpose on the fly; when we employ some automatic vectorization, it will most certainly be defeated by transposition.

Reply:
Done